Add MemberMatters support

rechner opened the pull request:
Currently, all access control decisions at Pawprint are being made using LDAP as our source of truth.  Keeping LDAP users synced and turning off access after a user's payment has lapsed is a manual process.

MemberMatters has direct support for access control using a websocket protocol, this PR adds support for emulating a MemberMatters access control device to Doorman.

Doorman will first consult the access list from Member Matters, and can optionally fall-back to querying a card in LDAP.

The access list from MemberMatters is cached locally in the file specified by `DOORMAN_MM_DATA_FILE`, so even for our instance hosted in the cloud, the doorman instance running locally will continue to work during an internet outage.

rechner committed:
WIP

rechner committed:
Working MemberMatters websocket client

rechner committed:
Update envrc example to be .env compatible

rechner committed:
pre-commit update and run

kataze commented on doorman/app.py:
```diff
@@ -44,35 +49,47 @@ def return_code_template(status):
-def lookup_card(card_number):
+def _lookup_mm(card_number: str) -> bool:
+    with open(MM_DATA_FILE) as f:
+        mm_data = json.load(f)
+    authorized_tags = mm_data.get("tags", [])
+    locked_out = mm_data.get("locked_out", False)
+    if locked_out:
+        app.logger.info(
+            "This MemberMatters device was set to locked_out by the server!"
+        )
+        return False
+
+    app.logger.debug(
+        f"Loaded {len(authorized_tags)} tags from MemberMatters cache at {MM_DATA_FILE}"
+    )
+    if card_number in authorized_tags:
+        app.logger.info(f"card_number: {card_number} is authorized by MemberMatters")
+        return True
+
+    app.logger.debug(f"card_number: {card_number} not authorized by MemberMatters")
+    return False
+
+
```
```suggestion
def _lookup_mm(card_number: str) -> bool:
    with open(MM_DATA_FILE) as f:
        mm_data = json.load(f)
    app.logger.debug(
        f"Loaded {len(authorized_tags)} tags from MemberMatters cache at {MM_DATA_FILE}"
    )
    authorized_tags = mm_data.get("tags", [])
    locked_out = mm_data.get("locked_out", False)
    if locked_out:
        app.logger.info(
            "This MemberMatters device was set to locked_out by the server!"
        )
        return False

    if card_number in authorized_tags:
        app.logger.info(f"card_number: {card_number} is authorized by MemberMatters")
        return True

    app.logger.debug(f"card_number: {card_number} not authorized by MemberMatters")
    return False


```

kataze replied:
just trying to move the loaded from cache message but the diff looks awful. Maybe I screwed up some whitespace in this plaintext comment editor lol

kataze commented on doorman/app.py:
```diff
@@ -93,6 +110,13 @@ def lookup_card(card_number):
+def lookup_card(card_number: str) -> bool:
+    ret = _lookup_mm(card_number)
+    if not ret:
+        ret = _lookup_ldap(card_number)
```
Will this fallback authorize people even if the device is "locked_out"? Is that what we want?

rechner replied:
The idea here is to use MM as the primary source of truth, but allow for a break-glass fallback lookup in LDAP.

If MM cache doesn't authorize your card, but LDAP does, it should let you in.

kaybarkbark replied:
Can we log or alert when this happens? If this occurs, it's likely a configuration error. We should be notified so we don't just start relying on it.

kataze commented on websocket_client.py:
```diff
@@ -0,0 +1,129 @@
+            websocket.send(json.dumps({"command": "pong"}))
+
+        elif command == "reboot":
+            # Not implemented (lol)
+            pass
```
lol

kataze commented on websocket_client.py:
```diff
@@ -0,0 +1,129 @@
+            # Not implemented (lol)
+            pass
+
+        elif command == "bump":
```
would "open" be more clear? Or is it reserved somehow

rechner replied:
We are limited here by the existing MemberMatters API

kataze reviewed (commented):
Lookin' hot 🔥 

I think this will help remove a lot of maintenance overhead now that people are pausing and restarting their memberships all the time. Couple thoughts, nothing crazy.

rechner committed:
Apply suggestions from code review

Co-authored-by: Kataze the Skunk <kataze@fursplosion.com>

rechner committed:
Add door_access protocol support via threading

- Add doorman/ipc.py with shared threading.Queue
- Push door_access events from Flask on RFID auth success
- Poll queue in websocket client, send door_access to MM server
- Add main.py entrypoint running Flask + WS client in one process
- Update Dockerfile to use main module with WEB_CONCURRENCY=1

rechner committed:
Update pre-commit, apply formatting

rechner committed:
Add LDAP notification, mypy type checking

rechner committed:
Rewrite as FastAPI app, add Sentry instrumentation

Replace Flask/requests/gunicorn with FastAPI/httpx/uvicorn. The
MemberMatters websocket client now runs in a daemon thread started
by the FastAPI lifespan, with a separate pump thread forwarding
door_access events to the server (drops the rel dispatcher, which
can't install signal handlers off the main thread).

Other fixes:
- Wrap blocking ldap3, file, and apprise calls in asyncio.to_thread
  so they don't stall the event loop.
- Catch FanvilParseError in auth() and add a global exception handler
  that ensures we always return the delicious XML it likes.
- Share a single httpx.AsyncClient on app.state with explicit timeouts
  for success/doorbell webhooks; guard against None DOORBELL_WEBHOOK.
- open_door() is synchronous so the websocket worker thread doesn't
  spin up a fresh event loop and httpx pool on every bump.
- Logging configuration is really pretty in a terminal now.

rechner committed:
add server ping timeout and cron to send keep-alives

rechner committed:
Glitchtip-specific sentry config

rechner committed:
Webhook and telegram notifications to best-effort

rechner commented:
Reworked to use FastAPI, asyncio, and reconfigured the webhook and apprise notifications to be best-effort in a background thread so it doesn't hold up the response back to the door controller.

rechner committed:
Update mypy dependencies list

rechner committed:
Add GHA to build docker container
